### PR TITLE
Feature: Pre-commit hooks with husky

### PR DESCRIPTION
**Pull request by jnurkka**
This adds Husky as dev dependency and uses husky to add a pre-commit hook that runs tests, linter and formats files with prettier

### CODE REVIEW

**Review by sergey-cliniserve — APPROVED** [review]
I'm generally against using git hooks for almost any purpose for the sole reason that commits become time-expensive. Also, the adoption of hooks can lead us down the slippery slope of introducing CI things there. Running tests in a hook is a good example.
Because of these reasons I'm more or less fine with using it for the package like this one, but strongly against of it spreading to bigger packages. And as for bigger packages we'll anyways need to build a CI, I would expect that CI also covers the functionality you built here hooks for.

**Comment by jnurkka** [comment]
> I'm generally against using git hooks for almost any purpose for the sole reason that commits become time-expensive. Also, the adoption of hooks can lead us down the slippery slope of introducing CI things there. Running tests in a hook is a good example.
> Because of these reasons I'm more or less fine with using it for the package like this one, but strongly against of it spreading to bigger packages. And as for bigger packages we'll anyways need to build a CI, I would expect that CI also covers the functionality you built here hooks for.

I agree that tests should only be run in a pre-commit hook as long as they take less than 10 seconds or something to run. However, I feel that formatting files that have been staged really makes sense, otherwise developers end up having to make commits for re-formatting the changes they already made. Or what do you think? And formatting or even linting files that were changed should be a very quick ordeal AFAIK.

**Comment by sergey-cliniserve** [comment]
> > I'm generally against using git hooks for almost any purpose for the sole reason that commits become time-expensive. Also, the adoption of hooks can lead us down the slippery slope of introducing CI things there. Running tests in a hook is a good example.
> > Because of these reasons I'm more or less fine with using it for the package like this one, but strongly against of it spreading to bigger packages. And as for bigger packages we'll anyways need to build a CI, I would expect that CI also covers the functionality you built here hooks for.
> 
> I agree that tests should only be run in a pre-commit hook as long as they take less than 10 seconds or something to run. However, I feel that formatting files that have been staged really makes sense, otherwise developers end up having to make commits for re-formatting the changes they already made. Or what do you think? And formatting or even linting files that were changed should be a very quick ordeal AFAIK.

How I understand the main goal we want to achieve is to reduce the size of formatting diffs in PRs. The right ground step we need to make is to disallow merging unformatted files. The only reliable way to achieve it is to block such PRs from merging and this can be done by CI checks only.
Once we have this, obviously all devs will be immediately annoyed by CI failing too often and there we need to make a second step of formatting routine automation. Options we so far came up with are:
1. hooks
2. IDE pre-save

IDE pre-save feature is already on for everyone, just some eslint/prettier configs are broken so it doesn't work reliably. And it already causes pretty much of annoyance when IDE hangs for 5-10 secs. However, this feature really speeds up writing the code, because without it you often end up adjusting the code manually.
Assuming pre-save works reliably I don't see what exactly hooks could help with. So we rather need to fix the configs used by IDEs.

Generally about hooks I'm worried that they are not made for this purpose. If we'd use git on a very basic level like `one feature = one commit`, then it'd be fine. But as we discussed many times on learning sessions git is really an amazing tool and can improve DX a lot if you learn it deeper. Commit's purpose is broader than fixating the code that's ready to be published. It often happens you commit some WIP work or rearrange existing commits during rebase so you just want to put the code to the git history and not apply any effects to it.
Configuring formatting as a hook breaks those advanced usage scenarios and makes git usage experience worse which can also lead to the development of a dislike for devs who are yet learning those advanced functionalities.

What I would use pre-commit hooks for is checks of commit name/message for example.

**Comment by jnurkka** [comment]
> How I understand the main goal we want to achieve is to reduce the size of formatting diffs in PRs. The right ground step we need to make is to disallow merging unformatted files. The only reliable way to achieve it is to block such PRs from merging and this can be done by CI checks only. Once we have this, obviously all devs will be immediately annoyed by CI failing too often and there we need to make a second step of formatting routine automation. Options we so far came up with are:
> 
> 1. hooks
> 2. IDE pre-save
> 
> IDE pre-save feature is already on for everyone, just some eslint/prettier configs are broken so it doesn't work reliably. And it already causes pretty much of annoyance when IDE hangs for 5-10 secs. However, this feature really speeds up writing the code, because without it you often end up adjusting the code manually. Assuming pre-save works reliably I don't see what exactly hooks could help with. So we rather need to fix the configs used by IDEs.

I think the problem is that we can't rely on pre-save working reliably. My feeling is that there will always be issues with VSCode plugins and we simply can't make sure that every develop always 100% uses the same config. I think it is simply way more reliable to allow CLI to take care of formatting before the files are committed.

I think another reason speaking for doing code formatting in a pre-commit hook is that we don't want developers to have to waste time being blocked by CI, manually going back to code and formatting it, adding one more commit and waiting for CI to finish running again to see if everything is fine now. We rather want formatting to be something always done automatically without the developer needing to do anything. Just like it would with _format on save_, if we could guarantee that works. And even if it works, formatting staged files in pre-commit hook should never really take a noticeable amount of time, my experience is that the prettier CLI is way faster than the slow VSCode format on save action for example.

> Generally about hooks I'm worried that they are not made for this purpose. If we'd use git on a very basic level like `one feature = one commit`, then it'd be fine. But as we discussed many times on learning sessions git is really an amazing tool and can improve DX a lot if you learn it deeper. Commit's purpose is broader than fixating the code that's ready to be published. It often happens you commit some WIP work or rearrange existing commits during rebase so you just want to put the code to the git history and not apply any effects to it. Configuring formatting as a hook breaks those advanced usage scenarios and makes git usage experience worse which can also lead to the development of a dislike for devs who are yet learning those advanced functionalities.
> 
> What I would use pre-commit hooks for is checks of commit name/message for example.

I totally agree with your point and will remove running tests from the git hooks. This is definitely something that should be part of CI only. About commit messages I agree, just would need clear rules for them first, but then I think we should add that to the pre-commit hook.

**Comment by jnurkka** [comment]
@sergey-cliniserve to address our discussion above, I added lint-staged dependency and now only lint and format staged files in the pre-commit hook. Is this approach fine for you?

I will add a CI action next that tests and lints all files, but as I described above, I think it would be great if we would not have to rely on format on save actions of IDEs as they don't see reliable enough to me. I also don't see the downsides of forcing formatting staged files on pre-commit, as it basically is just a safety mechanism and does not cost almost any time at all

**Comment by sergey-cliniserve** [comment]
> > How I understand the main goal we want to achieve is to reduce the size of formatting diffs in PRs. The right ground step we need to make is to disallow merging unformatted files. The only reliable way to achieve it is to block such PRs from merging and this can be done by CI checks only. Once we have this, obviously all devs will be immediately annoyed by CI failing too often and there we need to make a second step of formatting routine automation. Options we so far came up with are:
> > 
> > 1. hooks
> > 2. IDE pre-save
> > 
> > IDE pre-save feature is already on for everyone, just some eslint/prettier configs are broken so it doesn't work reliably. And it already causes pretty much of annoyance when IDE hangs for 5-10 secs. However, this feature really speeds up writing the code, because without it you often end up adjusting the code manually. Assuming pre-save works reliably I don't see what exactly hooks could help with. So we rather need to fix the configs used by IDEs.
> 
> I think the problem is that we can't rely on pre-save working reliably. My feeling is that there will always be issues with VSCode plugins and we simply can't make sure that every developer always 100% uses the same config. I think it is simply way more reliable to allow CLI to take care of formatting before the files are committed.

Why do you feel so? To me the issue is that we're changing eslint/prettier configs with absolutely no care and then it leads to those configs being incompatible and IDE pre-save feature not working. If we won't do that and fix existing configs then I think we shouldn't have any issues with pre-save. And we must fix configs anyways because otherwise DX in IDEs is horrible. It's valid for both VSCode and WebStorm. As I told it's a bit better in WebStorm but still not working from time to time.
Another problem is that some developer may have the pre-save off, but I don't think this is enough reason to decrease DX of the rest of the team.
Allowing CLI to take care of it is of course way more reliable, but don't you feel we're not fixing the root cause and instead just hiding the implications of it?

> I think another reason speaking for doing code formatting in a pre-commit hook is that we don't want developers to have to waste time being blocked by CI, manually going back to code and formatting it, adding one more commit and waiting for CI to finish running again to see if everything is fine now. We rather want formatting to be something always done automatically without the developer needing to do anything. Just like it would with _format on save_, if we could guarantee that works. And even if it works, formatting staged files in pre-commit hook should never really take a noticeable amount of time, my experience is that the prettier CLI is way faster than the slow VSCode format on save action for example.

Above I mentioned we need either hooks or pre-save, I'm of course against manual formatting.

> We rather want formatting to be something always done automatically without the developer needing to do anything.

yes, but we don't want formatting to happen on every commit or is it the case only for me? This way I lose confidence in the tool and have to check commits not only for the content I expected to end up there, but also for additional changes that I can't predict.

> > Generally about hooks I'm worried that they are not made for this purpose. If we'd use git on a very basic level like `one feature = one commit`, then it'd be fine. But as we discussed many times on learning sessions git is really an amazing tool and can improve DX a lot if you learn it deeper. Commit's purpose is broader than fixating the code that's ready to be published. It often happens you commit some WIP work or rearrange existing commits during rebase so you just want to put the code to the git history and not apply any effects to it. Configuring formatting as a hook breaks those advanced usage scenarios and makes git usage experience worse which can also lead to the development of a dislike for devs who are yet learning those advanced functionalities.
> > What I would use pre-commit hooks for is checks of commit name/message for example.
> 
> I totally agree with your point and will remove running tests from the git hooks. This is definitely something that should be part of CI only. About commit messages I agree, just would need clear rules for them first, but then I think we should add that to the pre-commit hook.

no no, of course I'm not proposing to implement checks for messages. I just wanted to give an example of what hooks could be safely used for.

To sum it up. My worry is that committing changes doesn't necessary mean you want changes to be linted. And by enabling hooks we limit the use of git commit. The issue lays in corrupted eslint/prettier configuration, not in IDE plugins or lack of automation. And hooks are not fixing configs, but just enable another automation.
We can try it if you still feel strongly about it and perhaps I'm just too pessimistic about this approach. Just disappointed that we're not fixing the real cause and won't fix it even longer if hooks will smoothen the DX a bit.
